csrc/scheduler/mma_utils.cpp

```diff
@@ -837,6 +837,117 @@ void WarpMmaSwizzler::scheduleLdMatrix(TensorView* tv, MmaOperand operand) {
   setWarpMapped(tv, 2);
 }
 
+void WarpMmaSwizzler::parallelizeAsBulkSkippingFirstIDs(
+    TensorView* tv,
+    size_t first_ids_to_skip) {
+  size_t skip = 0;
+  for (auto id : tv->getLoopDomain()) {
+    if (skip < first_ids_to_skip) {
+      skip++;
+      continue;
+    }
+    id->parallelize(ParallelType::Bulk);
+  }
+}
+
+void WarpMmaSwizzler::swizzleTMABox(
+    TensorView* tv,
+    MmaInputSmemSwizzle swizzle,
+    bool split_outer_dim) {
+  auto dtype = tv->getDataType().value();
+  // Input is on the form:
+  // [...., K (assume is 16), NI (16 .. say dtype is half and swizzle
+  // size is 32B, N could have already been split to create the TMA box)].
+  // Here the TMA box is [16,16]. If the outer dim was split then the above
+  // input would be [KO(2), KI(8), NI(16)] with the box being [8, 16]
+
+  // Split outer Dim:
+  // [..., KO(2), KI(8), NI (16)] ->
+  // [..., KO(2), KI0(2), KII(4), NI (16)]
+  // Not Split:
+  // [..., K(16),  NI(16)] ->
+  // [..., KO(4), KI(4), NI(16)]
+  tv->split(-2, (128 / (getBytesFromSwizzle(swizzle))));
+
+  // Split outer Dim:
+  // [..., KO(2), KI0(2), KII(4), NI (16)] ->
+  // [..., KO(2), KI0(2), KII(4), NIO(2) ,NII(8)]
+  // Not Split:
+  // [..., KO(4), KI(4), NI(16)] ->
+  // [..., KO(4), KI(4), NIO(2), NII(8)]
+  tv->split(-1, (core_matrix_width_bytes / dataTypeSize(dtype)));
+
+  if (!split_outer_dim) {
+    // [..., KO(4), KI(4), NIO(2), NII(8)] -> [..., KOO(2), KOI(2) KI(4),
+    // NIO(2), NII(8)]
+    tv->split(-4, (getBytesFromSwizzle(swizzle) / 16));
+  }
+
+  tv->swizzle(SwizzleType::XOR, -4, -2);
+}
+
+// Please note that we currently do not fully support
+// not splitting the outer dimension. This only works when
+// the inner-dimension is not split, that is the inner dim
+// is less or equal to the swizzle size (in bytes).
+void WarpMmaSwizzler::scheduleTMALoadForMma(
+    TensorView* tv,
+    MmaInputSmemSwizzle swizzle,
+    bool split_outer_dim) {
+  // In the comments below I have kept K as the outer dimension. That is
+  // just to have a concrete running example - it can be inner or outer.
+
+  auto num_ids_to_skip = tv->getLoopDomain().size() - 2;
+  NVF_ERROR(num_ids_to_skip >= 0);
+  if (swizzle == MmaInputSmemSwizzle::None) {
+    // For no-swizzle case, the entire tile are divided into 8x8 core matrices,
+    // and each core matrix resides in a contiguous 8*8*2 bytes region in shared
+    // memory. [K, N]
+    tv->split(-2, 8);
+    tv->split(-1, 8);
+    // [Ko, K8, No, N8]
+    tv->reorder({{-2, -3}});
+    // [Ko, No, K8, N8]
+    num_ids_to_skip += 2;
+  } else {
+    auto dtype = tv->getDataType().value();
+
+    // In the comments below I assume K=16, N=32, swizzle=32, dtype = half.
+    if (split_outer_dim) {
```

Comment:
Should we always do the split? This way, we will use this flag to control the reorder dim, instead of whether doing split.

Reply:
While implementing this I realized this may have some drawback. Say I tile in both cases and reorder based on the boolean flag then the IDs look like:

[... NO(2), KO(2), KI(8), NI(16] {not split}
[... KO(2), NO(2), KI(8), NI(16)] {split}

Now when we then call `swizzleTMABox`, in that function I would need to remove the bit of code:

```
  if (!split_outer_dim) {
    // [..., KO(4), KI(4), NIO(2), NII(8)] -> [..., KOO(2), KOI(2) KI(4),
    // NIO(2), NII(8)]
    split(-4, (getBytesFromSwizzle(swizzle) / 16));
  }
```

That would mean that `swizzleTMABox` would expect the outer dim to have been tiled by 8 by it's users. I am not sure this is always preferable, as it might be nice to have users call this on a TMA box directly without tiling. I don't have strong opinions about this.

Reply:
Can we move the code that tile 8 into `swizzleTMABox`? Is there a case that we must not tile 8?

Reply:
This is the schedule of tile 8 vs not tile 8:
<img width="703" alt="image" src="https://github.com/user-attachments/assets/da3ba161-1436-4134-9293-dd0e4fff15c2">
The only difference is the highlighted area. Even if the user didn't tile 8, we will tile 4 then 2, which according to Theorem 2.1 (Equivalence of Split-Split) in [The Mathematical Theory of IterDomain
](https://github.com/NVIDIA/Fuser/blob/main/doc/reading/iterdomain.md), this is equivalent to the hightlighted area when the user did tile 8. It sounds tedious to use a flag to switch between mathematically equivalent forms.